### PR TITLE
implemented a cache for already downloaded models

### LINKED ISSUE #37: Add a cache for already downloaded models
When we create a model we have multiple places where the original model (onnx) could be downloaded. Currently, these downloads happen again and again as there is no cache or we dont look if we already downloaded it.

Some functions that download a model:

* `_get_output_dtype`: used to get the output type so users do not have to add it into the predict method. This could be potentially executed in each prediction
* `_download_model`: if an `output_path` is added into the instantiation of the model or agent, the model is downloaded

Prior to download the model we should check if we have the model already downloaded locally.

### PR DESCRIPTION
**Pull request by shivam6862**
@raphaelDkhn 
Implemented a cache for already downloaded models by using the diskcache package.  
Closes #37

### CODE REVIEW

File: giza_actions/model.py [hunk 1]
```diff
@@ -163,9 +168,16 @@ def _set_session(self):
...
-                self.model.id, self.version.version
-            )
+            cache_str = f"{self.model.id}_{self.version.version}_model"
+            if cache_str in self.cache:
+                file_path = self.cache.get(cache_str)
+                file_path = Path(file_path)
+                with open(file_path, "rb") as f:
+                    onnx_model = f.read()
+            else:
+                onnx_model = self.version_client.download_original(
```
Comment (Gonmeso) [inline giza_actions/model.py:178]: Let's change this to use `_download_model` and the read it, so we have the benefits of the cache.

File: giza_actions/model.py [hunk 2]
```diff
@@ -396,9 +417,16 @@ def _get_output_dtype(self):
...
-            self.model.id, self.version.version
-        )
+        cache_str = f"{self.model.id}_{self.version.version}_model"
+        if cache_str in self.cache:
+            file_path = self.cache.get(cache_str)
+            file_path = Path(file_path)
+            with open(file_path, "rb") as f:
+                file = f.read()
+        else:
+            file = self.version_client.download_original(
```
Comment (Gonmeso) [inline giza_actions/model.py:427]: We should change this to use `_download_model` which handles the usage of the cache if it does not exist. Currently if this is used multiple times the model would be download each time as is not reflected on the cache.

Lets use `_download_model` and the read.

**Comment by shivam6862** [comment]
@Gonmeso requested changes have been made

File: pyproject.toml [hunk 3]
```diff
@@ -13,6 +13,7 @@ license = "MIT"
 
 [tool.poetry.dependencies]
 python = ">=3.11,<4.0"
+diskcache == "5.6.3"
```
[the remark this answers is not in the document]
Comment (shivam6862) [reply pyproject.toml]: will run  poetry add diskcache

Package operations: 47 installs, 1 update, 0 removals
  - Installing pendulum (2.1.2): Failed
  ChefBuildError
  
Note: This error originates from the build backend, and is likely not a problem with poetry but with pendulum (2.1.2) not supporting PEP 517 builds. You can verify this by running 'pip wheel --no-cache-dir --use-pep517 "pendulum (==2.1.2)"'.

File: giza_actions/model.py [hunk 4]
```diff
@@ -83,13 +81,14 @@ def __init__(
             self._get_credentials()
             self.model = self._get_model(id)
             self.version = self._get_version(version)
-            self.session = self._set_session()
             self.framework = self.version.framework
             self.uri = self._retrieve_uri()
             self.endpoint_id = self._get_endpoint_id()
+            if output_path:
+                self.session = self._set_session(output_path)
             if output_path:
```
Comment (Gonmeso) [inline giza_actions/model.py:89]: Having to pass `output_path` to `_download_model` and `_get_output_dtype` will mostly make `output_path` to be mandatory, but this is not the intention so another path where if `output_path` is not provided we handle it for the user:

* Lets make the class to have an attribute `self._output_path`
* If `output_path` is present when the user creates the instance, we will use that path, if not we will handle that
* Remove `output_path` argument from `_download_model` and use `self._output_path`
* Remove `output_path` from `_set_session` and `_get_output_dtype` as it is not needed any more

An example of the init could be:

```python
    def __init__(
        self,
        model_path: Optional[str] = None,
        id: Optional[int] = None,
        version: Optional[int] = None,
        output_path: Optional[str] = None,
    ):
       ...

        if model_path:
            self.session = ort.InferenceSession(model_path)
        elif id and version:
            self.model_id = id
            self.version_id = version
            self.model_client = ModelsClient(API_HOST)
            self.version_client = VersionsClient(API_HOST)
            self.api_client = ApiClient(API_HOST)
            self.endpoints_client = EndpointsClient(API_HOST)
            self._get_credentials()
            self.model = self._get_model(id)
            self.version = self._get_version(version)
            self.session = self._set_session()
            self.framework = self.version.framework
            self.uri = self._retrieve_uri()
            self.endpoint_id = self._get_endpoint_id()
            if output_path is not None:
                self._output_path = output_path
            else:
                self._output_path = os.path.join(tempfile.gettempdir(), f"{self.model_id}_{self.version_id}_{self.model.name})
            # Now this internally uses self._output_path
            # As we are using the cache hitting this function should not be problematic
            self._download_model()
```

File: giza_actions/model.py [hunk 5]
```diff
@@ -163,9 +167,13 @@ def _set_session(self):
             )
 
         try:
-            onnx_model = self.version_client.download_original(
-                self.model.id, self.version.version
-            )
+            cache_str = f"{self.model.id}_{self.version.version}_model"
```
Comment (Gonmeso) [inline giza_actions/model.py:170]: Let's remove this and use `self._output_path` so cache keys are more consistent

File: giza_actions/model.py [hunk 6]
```diff
@@ -163,9 +167,13 @@ def _set_session(self):
...
+            self._download_model(output_path)
+
+            if cache_str in self._cache:
```
Comment (Gonmeso) [inline giza_actions/model.py:173]: With the proposed changes
```python
          if self._output_path in self._cache:
                file_path = Path(self._cache.get(self._output_path))
                with open(file_path, "rb") as f:
                    onnx_model = f.read()
```

File: giza_actions/model.py [hunk 7]
```diff
@@ -189,21 +197,28 @@ def _download_model(self, output_path: str):
                 f"Model version status is not completed {self.version.status}"
             )
 
-        onnx_model = self.version_client.download_original(
-            self.model.id, self.version.version
-        )
+        cache_str = f"{self.model.id}_{self.version.version}_model"
```
Comment (Gonmeso) [inline giza_actions/model.py:200]: As in the previous comment, let's remove this and use self._output_path so cache keys are more consistent

File: giza_actions/model.py [hunk 8]
```diff
@@ -189,21 +197,28 @@ def _download_model(self, output_path: str):
...
+                self.model.id, self.version.version
+            )
 
-        if ".onnx" in output_path:
-            save_path = Path(output_path)
-        else:
-            save_path = Path(f"{output_path}/{self.model.name}.onnx")
+            logger.info("ONNX model is ready, downloading! ✅")
+
+            if ".onnx" in output_path:
```
Comment (Gonmeso) [inline giza_actions/model.py:209]: With the proposed changes this should be `self._output_path`

File: giza_actions/model.py [hunk 9]
```diff
@@ -221,6 +236,7 @@ def predict(
         custom_output_dtype: Optional[str] = None,
         job_size: str = "M",
         dry_run: bool = False,
+        output_path: Optional[str] = None,
```
Comment (Gonmeso) [inline giza_actions/model.py:239]: Remove this as it is not needed

File: giza_actions/model.py [hunk 10]
```diff
@@ -272,7 +288,7 @@ def predict(
                     logger.info("Serialized: %s", serialized_output)
 
                     if custom_output_dtype is None:
-                        output_dtype = self._get_output_dtype()
+                        output_dtype = self._get_output_dtype(output_path)
```
Comment (Gonmeso) [inline giza_actions/model.py:291]: This should be as it was before

File: giza_actions/model.py [hunk 11]
```diff
@@ -388,17 +404,21 @@ def _parse_cairo_response(self, response, data_type: str):
...
         Retrieve the Cairo output data type base on the operator type of the final node.
 
         Returns:
             The output dtype as a string.
         """
 
-        file = self.version_client.download_original(
-            self.model.id, self.version.version
-        )
+        cache_str = f"{self.model.id}_{self.version.version}_model"
```
Comment (Gonmeso) [inline giza_actions/model.py:415]: With the proposed changes this should be `self._output_path`

**Comment by Gonmeso** [comment]
Testing is still missing, that its why I proposed creating a cache class to test it easily.

**Comment by shivam6862** [comment]
@Gonmeso @raphaelDkhn  added a Test for cache and Requested changes have been made

File: giza_actions/model.py [hunk 12]
```diff
@@ -79,12 +82,19 @@ def __init__(
...
+            self.session = self._set_session()
+            if output_path is not None:
+                self._output_path = output_path
+            else:
+                self._output_path = os.path.join(
+                    tempfile.gettempdir(),
+                    f"{self.model_id}_{self.version_id}_{self.model.name}",
+                )
+            self._download_model()
+            self._cache = Cache(os.path.join(os.getcwd(), "tmp", "cachedir"))
```
Comment (Gonmeso) [inline giza_actions/model.py:97]: Cache should be initialized before `_set_session` and `download_model` as if the user does not provide `output_path` set session will try to hit `self._cache` which has not been initialized. 

This is making the previous two existing tests to fail.

File: tests/test_model.py [hunk 13]
```diff
@@ -102,3 +102,22 @@ def test_predict_success_with_file(*args):
 
     assert np.array_equal(result, expected)
     assert req_id == "123"
+
+
+@patch("giza_actions.model.GizaModel._set_session")
+@patch("giza_actions.model.GizaModel._get_output_dtype")
+@patch("giza_actions.model.GizaModel._retrieve_uri")
+@patch("giza_actions.model.GizaModel._get_endpoint_id", return_value=1)
+def test_cache_implementation(*args):
```
Comment (Gonmeso) [inline tests/test_model.py:111]: The test is failing as some mocks are missing, the current error is due to missing credentials, for that just add:

```python
@patch("giza_actions.model.GizaModel._get_credentials")
def test_cache_implementation(*args):
```

Also it is possible that some patches are missing as well, for example for `_get_model` and `_get_version`  might be need.

**Review by Gonmeso — CHANGES_REQUESTED** [review]
Really close to be done!

We just need to fix the tests and this would be ready, as CI is failing. Try to make sure that everything is working previously by running `pytest`

**Comment by shivam6862** [comment]
@Gonmeso Added Patch for test and made required changes to pass CI,
I'm looking for some guidance.

**Comment by Gonmeso** [comment]
Hi!

One little note, making the changes and just waiting CI to pass will make this process longer, it is encouraged to run the tests locally and then push the changes when they pass locally.

The error is the following:

```python
/home/runner/work/actions-sdk/actions-sdk/giza_actions/model.py:206: in _download_model
    onnx_model = self.version_client.download_original(
E   Exception: Token expired or not set. API Key not available. Log in again.
        self       = <giza_actions.model.GizaModel object at 0x7fcfd48[65](https://github.com/gizatechxyz/actions-sdk/actions/runs/8814392110/job/24195017336?pr=39#step:6:66)890>
```

Here we can see that `_download_model` is being executed and the next call in the stack is `self.version_client.download_original` which is trying to get the model from the API and fails to do so without credentials, that means that we need to patch this.

With the recent changes `_download_model`, it is being executed every single time as currently we always handle the output path to use with the cache.

To solve this we could start by patching the `self.version_client.download_original` function. Checking this function, this dependency comes in the `__init__` method:

```python
            self.model_client = ModelsClient(API_HOST)
            self.version_client = VersionsClient(API_HOST) # < This one 
            self.api_client = ApiClient(API_HOST)
            self.endpoints_client = EndpointsClient(API_HOST)
```

In order to patch this, we need to patch the imported client from the script (docs here: https://docs.python.org/3/library/unittest.mock.html#where-to-patch):

```python
@patch("giza_actions.model.VersionsClient", return_value=b"some bytes")
```

We add a returned value of bytes because it is what `self.version_client.download_original` returns.

Hope this helps @shivam6862

**Comment by shivam6862** [comment]
Screenshot 2024-04-24 200923

@Gonmeso add test case are passed

**Comment by shivam6862** [comment]
@Gonmeso done for all file  **run pre-commit run --all-files** which was done previously for **.gitignore**

**Review by Gonmeso — APPROVED** [review]
Thanks for your input!

LGTM!!!!

Great work @shivam6862 ⚡

**Pull request merged**